SqlClient: Reduce magic strings

justinvp opened the pull request:
Use `nameof()` and `[CallerMemberName]` to reduce magic strings in SqlClient. Also delete unused string constants.

Part of #7290.

cc: @saurabh500

justinvp commented on src/System.Data.SqlClient/src/Microsoft/SqlServer/Server/SqlDataRecord.cs:
```diff
@@ -649,7 +649,7 @@ public SqlDataRecord(params SqlMetaData[] metaData)
             if (null == metaData)
             {
-                throw ADP.ArgumentNull("metadata");
+                throw ADP.ArgumentNull(nameof(metaData));
```
Deliberate change to match the casing of the parameter.

justinvp commented on src/System.Data.SqlClient/src/Microsoft/SqlServer/Server/SqlDataRecord.cs:
```diff
@@ -658,7 +658,7 @@ public SqlDataRecord(params SqlMetaData[] metaData)
                 if (null == metaData[i])
                 {
-                    throw ADP.ArgumentNull("metadata[" + i + "]");
+                    throw ADP.ArgumentNull($"{nameof(metaData)}[{i}]");
```
Deliberate change to match the casing of the parameter.

justinvp commented on src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs:
```diff
@@ -827,69 +828,14 @@ static internal Exception TransactionZombied(DbTransaction obj)
-        internal const string UnquoteIdentifier = "UnquoteIdentifier";
-        internal const string Update = "Update";
-        internal const string UpdateCommand = "UpdateCommand";
-        internal const string UpdateRows = "UpdateRows";
```
All the removed constants above were either dead code before this PR or became dead code as part of this PR by switching uses to `nameof()` or `[CallerMemberName]`.

saurabh500 commented on src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs:
```diff
@@ -403,7 +404,7 @@ static internal InvalidOperationException NoConnectionString()
         }
 
-        static internal Exception MethodNotImplemented(string methodName)
+        static internal Exception MethodNotImplemented([CallerMemberName] string methodName = "")
```
@justinvp Why did you change these?

justinvp replied:
So callers can call it like `MethodNotImplemented();` instead of `MethodNotImplemented("Caller");` and the compiler will automatically fill in `"Caller"`.

saurabh500 replied:
Commented before looking at the rest of the Diff.

justinvp commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlStream.cs:
```diff
@@ -469,27 +469,27 @@ override public long Seek(long offset, SeekOrigin origin)
             if (null == _cachedBytes)
             {
-                throw ADP.StreamClosed(ADP.Read);
+                throw ADP.StreamClosed();
```
I deliberately changed this from the previously incorrect "Read" to the correct "Seek" (via `[CallerMemberName]`).

saurabh500 commented:
@justinvp Thanks for the changes.  I am going through them. 

cc @nadiatk @corivera @kkurni

stephentoub commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs:
```diff
@@ -889,13 +889,13 @@ private void BeginExecuteNonQueryInternalReadStage(TaskCompletionSource<object>
         }
 
-        private void VerifyEndExecuteState(Task completionTask, String endMethod)
+        private void VerifyEndExecuteState(Task asyncResult, String endMethod)
```
Seems a bit backwards to change the name of the parameter from the more applicable "completionTask" to the less applicable "asyncResult", simply to be able to use ```nameof(asyncResult)``` instead of ```"asyncResult"```.

justinvp replied:
`asyncResult` seems like just as reasonable a name as `completionTask`, in my opinion. `VerifyEndExecuteState` is only used in two places to verify the asyncResult, which happens to be a Task.

stephentoub replied:
It's strongly typed as a Task.  It's not like it's an IAsyncResult that happens to be a Task and is being cast to one internally.

saurabh500 replied:
I agree with @stephentoub on this. Otherwise the changes look good to me.

justinvp replied:
Since the public `Begin`/`End` methods don't exist in .NET Core, the `Task` passed in here should never be null, so I've replaced the `ArgumentNullException` with a `Debug.Assert` to avoid the parameter rename altogether.

stephentoub commented:
One nit, otherwise LGTM.

saurabh500 commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs:
```diff
@@ -748,7 +748,7 @@ override public object ExecuteScalar()
                 statistics = SqlStatistics.StartTimer(Statistics);
                 SqlDataReader ds;
-                ds = RunExecuteReader(0, RunBehavior.ReturnImmediately, true, ADP.ExecuteScalar);
+                ds = RunExecuteReader(0, RunBehavior.ReturnImmediately, true, nameof(ExecuteScalar));
```
Wouldn't it make more sense to change RunExecuteReader to take [CallerMemberName] string method=""
You can get rid of most of the nameof(...) being passed down to RunExecuteReader as well.

justinvp replied:
Good point. Updated.

justinvp commented:
Thanks for the feedback, @saurabh500 and @stephentoub. I added a commit that addresses your comments. (I temporarily added back a string constant to get rid of the public `EndExecuteNonQuery` method that is not in the public contract, to be reconciled in the subsequent commit).

I also added another commit that replaces the remaining string constants with uses of CallerMemberName and nameof, by removing the Async suffix from the private BeginXxxAsync and
EndXxxAsync methods. Please take a look.

justinvp commented:
Test Innerloop Windows_NT Debug Build and Test please

(#4467 and #7517)

justinvp commented:
Test Innerloop CentOS7.1 Release Build and Test please

stephentoub commented:
Thanks, Justin.
